### PR TITLE
Complete the remaining logic of WorkingSealerManager

### PR DESCRIPTION
**Pull request by cyjseagull**
1. Complete the remaining logic of WorkingSealerManager
2. Fix problems of VRFBasedrPBFT

### CODE REVIEW

File: libprecompiled/WorkingSealerManagerImpl.h [hunk 1]
```diff
@@ -77,19 +78,26 @@ class WorkingSealerManagerImpl
...
+        std::string const& _vrfInput);
     VRFInfo::Ptr vrfInfo() const { return m_vrfInfo; }
 
+    virtual void rotateWorkingSealer();
+
 protected:
     virtual void checkVRFInfos();
 
-    // calculate the number of working sealers that need to be added and removed
+    // calculate the number of working sealers that need to be removed and removed
```
Comment (bxq2011hust) [inline libprecompiled/WorkingSealerManagerImpl.h:90]: to be removed and removed ？

File: libconsensus/vrf_rpbft/VRFBasedrPBFTEngine.cpp [hunk 2]
```diff
@@ -32,7 +32,16 @@ void VRFBasedrPBFTEngine::setShouldRotateSealers(bool _shouldRotateSealers)
 // Working sealer is elected as leader in turn
 IDXTYPE VRFBasedrPBFTEngine::selectLeader() const
 {
-    return (IDXTYPE)((m_view + m_highestBlock.number()) % m_workingSealersNum);
+    int64_t selectedIdx = ((m_view + m_highestBlock.number()) % m_workingSealersNum);
+    dev::h512 selectedNodeId;
+    {
+        ReadGuard l(x_chosedSealerList);
```
[the remark this answers is not in the document]
Comment (cyjseagull) [reply libconsensus/vrf_rpbft/VRFBasedrPBFTEngine.cpp:38]: The member variables of the previous rPBFT are reused here~
I think chosed and chosen equivalent.

File: libconsensus/vrf_rpbft/VRFBasedrPBFTEngine.cpp [hunk 3]
```diff
@@ -83,6 +91,14 @@ void VRFBasedrPBFTEngine::updateConsensusNodeList()
     {
         ReadGuard l(m_sealerListMutex);
         m_sealersNum = m_sealerList.size();
+        WriteGuard lock(x_nodeID2Index);
```
[the remark this answers is not in the document]
Comment (cyjseagull) [reply libconsensus/vrf_rpbft/VRFBasedrPBFTEngine.cpp:94]: ok~ `x_nodeID2Index ` and `m_sealerListMutex` are used together only here.

File: libconsensus/vrf_rpbft/VRFBasedrPBFTEngine.h [hunk 4]
```diff
@@ -61,6 +62,9 @@ class VRFBasedrPBFTEngine : public dev::consensus::RotatingPBFTEngine
     // Used to notify Sealer whether it needs to rotate sealers
     std::atomic_bool m_shouldRotateSealers = {false};
     std::atomic<int64_t> m_workingSealersNum = {0};
+
+    std::shared_ptr<std::map<dev::h512, IDXTYPE>> m_nodeID2Index;
+    mutable SharedMutex x_nodeID2Index;
```
[the remark this answers is not in the document]
Comment (cyjseagull) [reply libconsensus/vrf_rpbft/VRFBasedrPBFTEngine.h:67]: According to the code specification, the lock is the `x_` prefix.

File: libprecompiled/WorkingSealerManagerImpl.h (hunk 1 above)
Comment (cyjseagull) [reply libprecompiled/WorkingSealerManagerImpl.h]: Modified.

**Pull request merged**